Linking liblibc using musl target fails ("unrecognized relocation") with the new nightlies

Here's a dockerized test case: https://github.com/golddranks/libc_regression_test/ (run with `docker build .`)

Here's the error:
```
  = note: /usr/bin/ld: /root/.rustup/toolchains/nightly-2017-02-19-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/liblibc-95af4192ed69a1c8.rlib(exit.lo): unrecognized relocation (0x2a) in section `.text.libc_exit_fini'
          /usr/bin/ld: final link failed: Bad value
          collect2: error: ld returned 1 exit status
```

The linking error seems to happen only with the musl target. I originally posted this as https://github.com/rust-lang/libc/issues/529, but because the breaking is dependent on the nightly, it seems that this is an `rustc` issue rather than an issue of the `libc` crate.

After testing with various nightlies, it seems that the nightly `nightly-2017-02-16 installed - rustc 1.17.0-nightly (62eb6056d 2017-02-15)` is the first one that breaks. The nightly before that, `nightly-2017-02-13 installed - rustc 1.17.0-nightly (956e2bcba 2017-02-12)` is the last one that works.

So that seems to imply that the commit that introduced this regression is in the range `956e2bcba..62eb6056d`. Any thoughts about which one it might be?

--- a/src/bootstrap/lib.rs
+++ b/src/bootstrap/lib.rs
@@ -828,17 +828,6 @@ impl Build {
         if target.contains("apple-darwin") {
             base.push("-stdlib=libc++".into());
         }
-        // This is a hack, because newer binutils broke things on some vms/distros
-        // (i.e., linking against unknown relocs disabled by the following flag)
-        // See: https://github.com/rust-lang/rust/issues/34978
-        match target {
-            "i586-unknown-linux-gnu" |
-            "i686-unknown-linux-musl" |
-            "x86_64-unknown-linux-musl" => {
-                base.push("-Wa,-mrelax-relocations=no".into());
-            },
-            _ => {},
-        }
         return base
     }
 

--- a/src/ci/docker/linux-tested-targets/Dockerfile
+++ b/src/ci/docker/linux-tested-targets/Dockerfile
@@ -34,6 +34,14 @@ ENV RUST_CONFIGURE_ARGS \
       --musl-root-x86_64=/musl-x86_64 \
       --musl-root-i686=/musl-i686
 
+# Newer binutils broke things on some vms/distros (i.e., linking against
+# unknown relocs disabled by the following flag), so we need to go out of our
+# way to produce "super compatible" binaries.
+#
+# See: https://github.com/rust-lang/rust/issues/34978
+ENV CFLAGS_i686_unknown_linux_gnu=-Wa,-mrelax-relocations=no \
+    CFLAGS_x86_64_unknown_linux_gnu=-Wa,-mrelax-relocations=no
+
 ENV SCRIPT \
       python2.7 ../x.py test \
           --target x86_64-unknown-linux-musl \

--- a/src/ci/docker/linux-tested-targets/build-musl.sh
+++ b/src/ci/docker/linux-tested-targets/build-musl.sh
@@ -11,7 +11,10 @@
 
 set -ex
 
-export CFLAGS="-fPIC"
+# We need to mitigate rust-lang/rust#34978 when compiling musl itself as well
+export CFLAGS="-fPIC -Wa,-mrelax-relocations=no"
+export CXXFLAGS="-Wa,-mrelax-relocations=no"
+
 MUSL=musl-1.1.14
 curl https://www.musl-libc.org/releases/$MUSL.tar.gz | tar xzf -
 cd $MUSL

